src/app-layer-dhcp.h

```diff
@@ -0,0 +1,180 @@
+#define DHCP_OPT_IP_REBINDING_TIME 59
+#define DHCP_OPT_VENDOR_CLASS 60
+#define DHCP_OPT_CLIENT_ID 61
+#define DHCP_OPT_TFTP_IP 66
+#define DHCP_OPT_END 255
+
+#define DHCP_PARAM_SUBNET_MASK 1
+#define DHCP_PARAM_ROUTER 3
+#define DHCP_PARAM_DNS_SERVER 6
+#define DHCP_PARAM_DOMAIN 15
+#define DHCP_PARAM_ARP_TIMEOUT 35
+#define DHCP_PARAM_NTP_SERVER 42
+#define DHCP_PARAM_TFTP_SERVER_NAME 66
+#define DHCP_PARAM_TFTP_SERVER_IP 150
+
+struct DHCPState_;
+
+typedef struct DHCPOpt_ {
+    uint8_t code;
+    uint8_t len;
+    uint8_t args[];
+} DHCPOpt;
+
+typedef struct DHCPTransaction_ {
+
+    struct DHCPState_ *state;
+
+    uint64_t tx_id;             /*<< Internal transaction ID. */
+
+    AppLayerDecoderEvents *decoder_events; /*<< Application layer
+                                            * events that occurred
+                                            * while parsing this
+                                            * transaction. */
+
+    uint32_t xid;
+    uint32_t logged;
+
+    uint8_t *request_buffer;
+    uint32_t request_buffer_len;
+
+    union {
+        uint32_t request_client_ip;
+        uint8_t request_client_ip_bytes[sizeof(uint32_t)];
+    };
+
+    uint8_t *response_buffer;
+    uint32_t response_buffer_len;
+
+    union {
+        uint32_t response_client_ip;
+        uint8_t response_client_ip_bytes[sizeof(uint32_t)];
+    };
+
+    uint32_t request_seen : 1; /*<< Flag to be set when the request is
+                                * seen. */
+    uint32_t response_seen : 1; /*<< Flag to be set when the response is
+                                * seen. */
+    uint32_t response_unneeded : 1;
+
+    DetectEngineState *de_state;
+
+    uint8_t reverse_flow; /*<< Set when the flow is the reverse of
+                           * what Suricata detected. This is because
+                           * the response can be on a new flow, which
+                           * Suricata will flag as to server. */
+
+    TAILQ_ENTRY(DHCPTransaction_) next;
+
+} DHCPTransaction;
+
+typedef struct DHCPGlobalState_ {
+
+    SCMutex lock;    /** Mutex for access to tx_list */
+
+
+    SC_ATOMIC_DECLARE(uint32_t, initialized);
+
+    TAILQ_HEAD(, DHCPTransaction_) tx_list; /**< List of DHCP transactions
+                                       * associated with this
+                                       * state. */
+
+    uint64_t transaction_max; /**< A count of the number of
+                               * transactions created.  The
+                               * transaction ID for each transaction
+                               * is allocted by incrementing this
+                               * value. */
+
+    /* Indicates the current transaction being logged. 
+     */
+    uint64_t log_id;
+
+    uint32_t transaction_count; /**< A count of the number of
+                                 * in-progress transactions. */
+
+    uint16_t events; /**< Number of application layer events created
```

Comment:
would it make sense to move this to DHCPState? When detection is done that should be the correct state right?